### PR TITLE
🦞 Expose Logits from inference for AUROC(=AUC) measurement

### PR DESCRIPTION
**Pull request by trajepl**
## Describe your changes
Current AUC computing is wrong which need the probabilities or logits for different classes.

This PR is used to expose Logits from inference for AUROC(=AUC) measurement

## Checklist before requesting a review
- [ ] Add unit tests for this change.
- [ ] Make sure all tests can pass.
- [ ] Update documents if necessary.
- [ ] Format your code by running `pre-commit run --all-files`
- [ ] Is this a user-facing change? If yes, give a description of this change to be included in the release notes.

### CODE REVIEW

File: olive/evaluator/accuracy.py [hunk 1]
```diff
@@ -114,28 +114,28 @@ def measure(self, preds, target):
...
 
-class AUC(AccuracyBase):
-    name: str = "auc"
+class AUROC(AccuracyBase):
+    name: str = "auroc"
 
-    def measure(self, preds, target):
-        preds_tensor, target_tensor = self.prepare_tensors(preds, target)
-        preds_tensor = preds_tensor.flatten()
+    def measure(self, preds, target, logits=None):
```
Comment (guotuofeng) [inline olive/evaluator/accuracy.py:127]: preds are useless?

File: olive/evaluator/olive_evaluator.py [hunk 2]
```diff
@@ -337,9 +338,11 @@ def _inference(
             # keep as numpy or torch arrays
             preds.append(outputs.cpu())
             targets.append(labels.cpu())
+            logits.append(result.cpu())
         preds = torch.cat(preds, dim=0)
         targets = torch.cat(targets, dim=0)
-        return preds, targets
+        logits = torch.cat(logits, dim=0)
+        return preds, targets, logits
```
Comment (guotuofeng) [inline olive/evaluator/olive_evaluator.py:345]: do we need define a class to contains all of them? is it possible that we need add other values in future?

File: olive/evaluator/olive_evaluator.py [hunk 3]
```diff
@@ -614,25 +620,30 @@ def _inference(
...
         targets = []
+        logits = []
         device = PyTorchEvaluator._device_string_to_torch_device(device)
         if device:
             session.to(device)
         for input_data, labels in dataloader:
             input_data = tensor_data_to_device(input_data, device)
             result = session(**input_data) if isinstance(input_data, dict) else session(input_data)
+            if not isinstance(result, torch.Tensor) and getattr(result, "logits", None) is not None:
+                result = result.logits
```
Comment (jambayk) [inline olive/evaluator/olive_evaluator.py:631]: a lot of our post funcs are already implement assuming the original state of `result`. I think it might be better to save the logit separately and not overwrite the original result.

File: olive/evaluator/olive_evaluator.py [hunk 4]
```diff
@@ -337,9 +338,11 @@ def _inference(
             # keep as numpy or torch arrays
             preds.append(outputs.cpu())
             targets.append(labels.cpu())
+            logits.append(result.cpu())
```
Comment (jambayk) [inline olive/evaluator/olive_evaluator.py:341]: result is not a torch tensor if the `len(output_names) >1`. It is going to be a dictionary of tensors.

File: olive/evaluator/olive_evaluator.py (hunk 3 above)
Comment (jambayk) [reply olive/evaluator/olive_evaluator.py:631]: This condition can be used at line 637 directly.

File: olive/evaluator/olive_evaluator.py (hunk 4 above)
Comment (guotuofeng) [reply olive/evaluator/olive_evaluator.py:341]: seconds this comment

File: olive/evaluator/olive_evaluator.py (hunk 2 above)
Comment (guotuofeng) [reply olive/evaluator/olive_evaluator.py:345]: something like https://huggingface.co/docs/transformers/main/main_classes/output#transformers.utils.ModelOutput

Comment (guotuofeng) [reply olive/evaluator/olive_evaluator.py:345]: is this line impacted? https://github.com/microsoft/Olive/pull/460/files#diff-816af95cf16e31610395bb20947f1d73cc1ca67355964e074a8f55375c8139a7R123

Comment (guotuofeng) [reply olive/evaluator/olive_evaluator.py:345]: ditto https://github.com/microsoft/Olive/blob/main/test/unit_test/assets/user_script.py#L8C5-L8C16

File: olive/evaluator/accuracy.py (hunk 1 above)
Comment (trajepl) [reply olive/evaluator/accuracy.py]: Updated with tuple(preds, logits)

File: olive/evaluator/olive_evaluator.py (hunk 2 above)
Comment (trajepl) [reply olive/evaluator/olive_evaluator.py]: Updated with tuple(preds, logits)

File: olive/evaluator/olive_evaluator.py (hunk 4 above)
Comment (trajepl) [reply olive/evaluator/olive_evaluator.py]: Updated.

File: olive/evaluator/accuracy.py [hunk 5]
```diff
@@ -77,15 +77,15 @@ def prepare_tensors(preds, target, dtypes=torch.int):
...
         raise NotImplementedError()
 
 
 class AccuracyScore(AccuracyBase):
     name: str = "accuracy_score"
 
-    def measure(self, preds, target):
-        preds_tensor, target_tensor = self.prepare_tensors(preds, target)
+    def measure(self, model_output, target):
+        preds_tensor, target_tensor = self.prepare_tensors(model_output[0], target)
```
Comment (jambayk) [inline olive/evaluator/accuracy.py:88]: I think it is clearer if we access it using `model_outputs.preds`

Comment (trajepl) [reply olive/evaluator/accuracy.py:88]: The main concern is about compatibility on tuple and named tuple. Once user want use Olive's evaluator, there is no need to pack the (preds, logits) with pre-defined name.

Comment (trajepl) [reply olive/evaluator/accuracy.py]: But it may only appear on the case user rewrite the inference function which is not supported in Olive now. 

Updated with named tuple.

**Review by guotuofeng — APPROVED** [review]
